### DISCUSSION TITLE
Best practices for design patterns

### DISCUSSION THREAD

**Discussion by AakashKumarNain** [jax-ml/jax#20497] [General]
I have dabbled with a lot of libraries and frameworks for NNs in the past. Given the pace of progress in ML, one library/framework isn't enough that suits all the needs. Some are mature, some have got the syntax and the design right, some have the momentum from the ecosystem. One of the most frustrating experience in all this is the design inconsistency between the frameworks/libraries
and the fragmentation within the ecosystem of a certain framework/library.

We all have been trying to develop something in JAX that closely match the flow in TF/PyTorch but also allows us to use the goodies that comes with the functional programming paradigm. Looking closely, we can split the design patters of all frameworks/libraries that are built on top of JAX into three categories:

1. `init-apply` pattern: Flax, Haiku, etc.
2. Pytrees based: Equinox, Haliax, etc.
3. Mix of OOP-stateless pattern as done [here](https://github.com/keras-team/keras/blob/97b082dfee2552fcad1a7c7ea0fac9c72943360c/keras/backend/common/stateless_scope.py) and [here](https://github.com/danijar/ninjax)

And there are other ways as well to build something on top of JAX, and it is extremely hard for anyone to figure out the best design pattern without trying them for a lot of uses or building something simple from scratch. What is the recommended design pattern if we have to develop something minimal (limited to NNs only), especially when we have to deal with both stateful and stateless operations?

**Comment by jakevdp** [comment]
I think best practices for NN APIs are still an active area of work, which is why you find different projects making different choices. Maybe in a few years the dust will have settled and there will be an answer to your question.

**Reply by AakashKumarNain** [reply]
Thanks for the response.

Fair enough but out of the above options, which one would you choose today? I am more interested in knowing the one that the JAX team thinks aligns with the JAX philosophy currently

**Reply by jakevdp** [reply]
I don't think there's any team-wide consensus here. I know @dougalm and @cgarciae have been thinking about these things recently; perhaps they could weigh in.

**Comment by cgarciae** [comment]
As Jake stated there is no consensus yet. Flax is the most widely used by both Google and HugginFace (apart from Keras) so it would be the safer option.

Since you are looking for "something in JAX that closely match the flow in TF/PyTorch" take a look at Flax's experimental [NNX](https://flax.readthedocs.io/en/latest/experimental/nnx/index.html) API. It represents Modules as "PyGraphs" and is able to express changes to the graph structure through jitted functions so model code tends to looks very similar to PyTorch/Keras:

```python
from flax.experimental import nnx

class Linear(nnx.Module):
  def __init__(self, din: int, dout: int, *, rngs: nnx.Rngs):
    key = rngs() # get a unique random key
    self.w = nnx.Param(jax.random.uniform(key, (din, dout)))
    self.b = nnx.Param(jnp.zeros((dout,))) # initialize parameters
    self.din, self.dout = din, dout

  def __call__(self, x: jax.Array):
    return x @ self.w.value + self.b.value

rngs = nnx.Rngs(0) # explicit RNG handling
model = Linear(din=2, dout=3, rngs=rngs) # initialize the model

x = jnp.empty((1, 2)) # generate random data
y = model(x) # forward pass
```

Stateful updates to Module in a jitted function are written as regular stateful python code:

```python
@nnx.jit
def add_one_to_bias(model: Linear):
  model.b.value += 1
```

**Reply by AakashKumarNain** [reply]
Thanks @cgarciae for the detailed response.

>Flax is the most widely used by both Google and HugginFace (apart from Keras) so it would be the safer option.

I think you misunderstood me. If I have to use something that is widely adopted and is sane, I would put my bets on Keras and Equinox for `n` number of reasons but my questions was not related to the widespread usage. My question is that given the evolution of these frameworks, does the JAX team believe that it still is the best approach to build on top of JAX? 

Re nnx: Can you please elaborate how it handles the stateful vs stateless behavior, say for example in BN?

**Reply by cgarciae** [reply]
I cannot comment for all of the JAX team as there are other designs being researched and I'm very biased towards Flax and NNX.

Regarding NNX, to handle stateful computation NNX embraces reference semantics by representing Modules as graphs instead of trees, by doing some bookkeeping `nnx.jit` is able to automatically propagate state updates. This is how BatchNorm looks in NNX:

```python
import jax
from flax.experimental import nnx

rngs = nnx.Rngs(0)
bn = nnx.BatchNorm(4, use_running_average=False, rngs=rngs)
print('BEFORE:', bn.get_state())

@nnx.jit
def forward(bn, x):
  return bn(x)

x = jax.random.uniform(rngs(), (10, 4))
y = forward(bn, x)
print('AFTER:', bn.get_state())
```
```
BEFORE: State({
  'bias': Param(
    raw_value=Array([0., 0., 0., 0.], dtype=float32)
  ),
  'mean': BatchStat(
    raw_value=Array([0., 0., 0., 0.], dtype=float32)
  ),
  'scale': Param(
    raw_value=Array([1., 1., 1., 1.], dtype=float32)
  ),
  'var': BatchStat(
    raw_value=Array([1., 1., 1., 1.], dtype=float32)
  )
})
AFTER: State({
  'bias': Param(
    raw_value=Array([0., 0., 0., 0.], dtype=float32)
  ),
  'mean': BatchStat(
    raw_value=Array([0.00364881, 0.0056361 , 0.00521917, 0.00481545], dtype=float32)
  ),
  'scale': Param(
    raw_value=Array([1., 1., 1., 1.], dtype=float32)
  ),
  'var': BatchStat(
    raw_value=Array([0.9904287 , 0.99087155, 0.9904572 , 0.99065596], dtype=float32)
  )
})
```

**Reply by adam-hartshorne** [reply]
Is there any significant additional overhead of using this stateful functionality?

I presume that because nnx make use of this bespoke PyGraph format, this makes interoperability with the likes of Equinox module, which keeps everything as a PyTree (which is native to JAX), difficult / impossible?

**Reply by AakashKumarNain** [reply]
> by doing some bookkeeping nnx.jit is able to automatically propagate state updates.

@cgarciae does this mean we cannot use `jax.jit` directly on nnx modules?

**Reply by cgarciae** [reply]
> Is there any significant additional overhead of using this stateful functionality?

@adam-hartshorne there is indeed an overhead but its roughly the same you get in Equinox (see [Equinox: Low-overhead training loops](https://docs.kidger.site/equinox/tricks/#low-overhead-training-loops)) and its just due to the fact that Python is slower that C/C++. To get around this overhead you can use a similar pattern to the one shown in the link, we can rewrite the above using NNX's [Functional API](https://flax.readthedocs.io/en/latest/experimental/nnx/nnx_basics.html#the-functional-api): 

```python
import jax
from flax.experimental import nnx

rngs = nnx.Rngs(0)
bn = nnx.BatchNorm(4, use_running_average=False, rngs=rngs)
static, state = nnx.split(bn)

@jax.jit
def forward(state, x):
  bn = static.merge(state)
  y = bn(x)
  state, _ = nnx.split(bn)
  return y, state

x = jax.random.uniform(rngs(), (10, 4))
y, state = forward(state, x)
```
Now the overhead is gone and we can also use regular `jax.jit`, but we have to do the bookkeeping ourselves.

**Reply by AakashKumarNain** [reply]
Nice! Thanks

**Reply by cgarciae** [reply]
@AakashKumarNain you cannot, but `nnx.jit` is a superset of `jax.jit` so you don't have to learn a new API, it just automates the `split` / `merge` shown above (more or less). You can also register your own types into NNX's graph system to teach NNX how to handle them similar how to register custom Pytree types, also Pytrees are automatically supported by NNX (as trees only).

**Reply by AakashKumarNain** [reply]
@cgarciae this is where my concern lies. If we have to wrap jit, that means there would be other transformations that would need to be wrapped around in nnx to make them work, right?

**Reply by cgarciae** [reply]
Correct. NNX provides versions for `jit`, `grad`, `vmap`, `scan`, and `remat`. However you can always use the [Functional API](https://flax.readthedocs.io/en/latest/experimental/nnx/nnx_basics.html#the-functional-api) to manually lift state into a transform we currently don't support or if you want more control. Ideally we support all transforms (with time) so this wouldn't be an issue.

For simple transforms that don't require semantics (e.g. `cond`) we could provide a simple combinator e.g:

```python
graph_cond = nnx.graph_transform(jax.lax.cond)
```

**Reply by AakashKumarNain** [reply]
Thanks but I think this provides enough support for my original question. Given that Equinox implements some of these, and nnx have to implement these, that basically means anyone who wishes to build on top of JAX have to implement the wrapped version of these transformations. I am not sure there will be a solution around it anytime soon, but in my opinion this is a friction

**Reply by ASEM000** [reply]
Also, For scientific computing, lifted transforms pattern sometimes fails with some libraries like jaxopt, so you end up rewriting you code to readapt to the split/merge pattern.

**Reply by AakashKumarNain** [reply]
Exactly! This is why I don't think this strategy of providing custom implementations will work out.
